### DISCUSSION TITLE
Updating file column by replacing the previous photo

### DISCUSSION THREAD

**Discussion by aleBranching** [pocketbase/pocketbase#5298] [Q&A] [answered]
Hello,
loving this project. Hoping I can contribute some day. 

I have a form that after submission could be edited. I use react hook form that I fill in the form with the data from pocketbase. 

This is how I create my form data

```javascript


      let formData = new FormData();
      Object.keys(data).forEach((key) => {
        if (key != "test_pattern_photo" && key != "if_no_fault_photo") {
          formData.append(key, data[key]);
        }
      });

      for (let i = 0; i < data.test_pattern_photo.length; i++) {
        formData.append("test_pattern_photo", data.test_pattern_photo[i]);
      }
      if (data.if_no_fault_photo != undefined) {
        for (let i = 0; i < data.if_no_fault_photo.length; i++) {
          formData.append("if_no_fault_photo", data.if_no_fault_photo[i]);
        }
      }
```

and this is how I update the file at the moment:

``` javascript

        if (formData.get("test_pattern_photo").constructor === File) {
          if (formData.get("if_no_fault_photo").constructor === File) {
            const record = await pb
              .collection("initial_questions")
              .update(previewData.id, {
                test_pattern_photo: "",
                if_no_fault_photo: "",
              });
          } else {
            const record = await pb
              .collection("initial_questions")
              .update(previewData.id, {
                test_pattern_photo: "",
              });
          }
        }

        const record = await pb
          .collection("initial_questions")
          .update(previewData.id, formData);

```

Calling update with `""` causes the picture to be deleted. Updating with just an image causes an append to the backend. 
The reason for checking if it's a `File` at `  if (formData.get("test_pattern_photo").constructor === File) ` is that pocketbase returns an array of filename strings. So if it's a string I know the image is not updated.

Is there a better way to replace an image? 

Also if I wanted to look at the source code to see how this file update is handled how would I find it by myself in the codebase?
Thank you.

**Comment by ganigeorgiev** [comment, marked as answer]
> ... So if it's a string I know the image is not updated.

Without any context about your collections structure and fields it is difficult to understand what is the desired behavior. The string value of the record field response indicates that the file was successfully uploaded (the string is the filename and it is what we store in the database).

As mentioned in the blue note in "[Uploading files](https://pocketbase.io/docs/files-handling/#uploading-files)",  if your field is a **single** `file` field, then uploading a new file will automatically replace the old one. If it is a **multiple** `file` field, then newly uploaded file(s) are appended to the existing array.

If you want to delete individual files from a **multiple** `file` field you can use the `-` modifier and specify the individual filename(s) to delete (e.g. `'fileField-': ['file1.txt', 'file2.txt']`).

_Note that with the [refactoring](https://github.com/pocketbase/pocketbase/discussions/4355), the files handling has slightly changed and for consistency with other multiple valued fields (`relation`, `select`) prepending/appending a new file is done only when the `+` modifier is used:_

```js
// file field modifiers and setters that will be supported with the upcoming v0.23.0
'fileField':  [file1, file2, ...]         // (changed) direct replace the old value with the new one
'+fileField': [file1, file2, ...]         // (new) prepend the new files to the existing array
'fileField+': [file1, file2, ...]         // (new) append the new files to the existing array
'fileField-': [filename1, filename2, ...] // delete the specified files from the existing array
```

If the above doesn't help, please elaborate a little more on your use case to understand the bigger picture and what actually you are trying to do.

---

> Also if I wanted to look at the source code to see how this file update is handled how would I find it by myself in the codebase?

The file upload/delete as part of an update request happen in the `forms.RecordUpsert` and more specifically:
- https://github.com/pocketbase/pocketbase/blob/238eddc4e028ce748e088d405987febda371aa02/forms/record_upsert.go#L157
- https://github.com/pocketbase/pocketbase/blob/238eddc4e028ce748e088d405987febda371aa02/forms/record_upsert.go#L822

But as mentioned earlier, with the refactoring this is done differently and the record files handling has changed significantly in attempt to make it easier to work with files programmatically when using PocketBase as Go framework.

**Comment by aleBranching** [comment]
Sorry should have removed irrelevant info in my question. My data is a multiple file field. 

> If you want to delete individual files from a multiple file field you can use the - modifier and specify the individual filename(s) to delete (e.g. 'fileField-': ['file1.txt', 'file2.txt']).

That's great I will be using that feature later. 

> 'fileField':  [file1, file2, ...]         // (changed) direct replace the old value with the new one

Aah yes that would be what I'm looking for. I will wait for v 0.23
For now I will do. 

``` javascript
            const record = await pb
              .collection("initial_questions")
              .update(previewData.id, {
                test_pattern_photo: "",
              });
```
Followed by 

```javascript
   const record = await pb
          .collection("initial_questions")
          .update(previewData.id, formData);
```
             
Thank you for the speedy response

**Reply by ganigeorgiev** [reply]
~You don't need 2 requests. You should be able to do the same with the existing APIs by doing something like:~

<details>
<summary>Expand</summary>

```js
'test_pattern_photo': null,
'test_pattern_photo+': yourNewFiles // edit: "+" file modifier is currently not supported
```
</details>

---

Sorry, ignore the above. I forgot that the file `+` append modifier is not available in the current version (see the list of the modifiers in https://pocketbase.io/docs/collections/).

In any case to avoid the double update request, you can try marking all existing files for deletion:

```js
'test_pattern_photo-': previewData.test_pattern_photo,
'test_pattern_photo': yourNewFiles,
```
